File: Tween/Foursquare.vb
Imports System.Net
Imports System.Runtime.Serialization
Imports System.Text.RegularExpressions

Public Class Foursquare
    Inherits HttpConnection

    Private Shared _instance As New Foursquare
    Public Shared ReadOnly Property GetInstance As Foursquare
        Get
            Return _instance
        End Get
    End Property

    Private _authKey As New Dictionary(Of String, String) From {
        {"client_id", "VWVC5NMXB1T5HKOYAKARCXKZDOHDNYSRLEMDDQYJNSJL2SUU"},
        {"client_secret", DecryptString("eXXMGYXZyuDxz/lJ9nLApihoUeEGXNLEO0ZDCAczvwdKgGRExZl1Xyac/ezNTwHFOLUZqaA8tbA=")}
    }

    Private CheckInUrlsVenueCollection As New Dictionary(Of String, Google.GlobalLocation)

    Private Function GetVenueId(ByVal url As String) As String
        Dim content As String = ""
        Try
            Dim res As HttpStatusCode = GetContent("GET", New Uri(url), Nothing, content)
            If res <> HttpStatusCode.OK Then Return ""
        Catch ex As Exception
            Return ""
        End Try
        Dim mc As Match = Regex.Match(content, "/venue/(?<venueId>[0-9]+)", RegexOptions.IgnoreCase)
        If mc.Success Then
            Dim vId As String = mc.Result("${venueId}")
            Return vId
        Else
            Return ""
        End If
    End Function

    Private Function GetVenueInfo(ByVal venueId As String) As FourSquareDataModel.Venue
        Dim content As String = ""
        Try
            Dim res As HttpStatusCode = GetContent("GET",
                                                   New Uri("https://api.foursquare.com/v2/venues/" + venueId),
                                                   _authKey,
                                                   content)

            If res = HttpStatusCode.OK Then
                Dim curData As FourSquareDataModel.FourSquareData = Nothing
                Try
                    curData = CreateDataFromJson(Of FourSquareDataModel.FourSquareData)(content)
                Catch ex As Exception
                    Return Nothing
                End Try

                Return curData.Response.Venue
            Else
                'Dim curData As FourSquareDataModel.FourSquareData = Nothing
                'Try
                '    curData = CreateDataFromJson(Of FourSquareDataModel.FourSquareData)(content)
                'Catch ex As Exception
                '    Return Nothing
                'End Try
                'MessageBox.Show(res.ToString + Environment.NewLine + curData.Meta.ErrorType + Environment.NewLine + curData.Meta.ErrorDetail)
                Return Nothing
            End If
        Catch ex As Exception
            Return Nothing
        End Try
    End Function

    Public Function GetMapsUri(ByVal url As String, ByRef refText As String) As String
        If Not AppendSettingDialog.Instance.IsPreviewFoursquare Then Return Nothing

        Dim urlId As String = Regex.Replace(url, "https?://(4sq|foursquare)\.com/", "")

        If CheckInUrlsVenueCollection.ContainsKey(urlId) Then
            refText = CheckInUrlsVenueCollection(urlId).LocateInfo
            Return (New Google).CreateGoogleStaticMapsUri(CheckInUrlsVenueCollection(urlId))
        End If

        Dim curVenue As FourSquareDataModel.Venue = Nothing
        Dim venueId As String = GetVenueId(url)
        If String.IsNullOrEmpty(venueId) Then Return Nothing

        curVenue = GetVenueInfo(venueId)
        If curVenue Is Nothing Then Return Nothing

        Dim curLocation As New Google.GlobalLocation With {.Latitude = curVenue.Location.Latitude, .Longitude = curVenue.Location.Longitude, .LocateInfo = CreateVenueInfoText(curVenue)}
        '例外発生の場合があるため
        If Not CheckInUrlsVenueCollection.ContainsKey(urlId) Then CheckInUrlsVenueCollection.Add(urlId, curLocation)
        refText = curLocation.LocateInfo
        Return (New Google).CreateGoogleStaticMapsUri(curLocation)
    End Function

    Private ReadOnly Property CreateVenueInfoText(ByVal info As FourSquareDataModel.Venue) As String
        Get
            Return info.Name + Environment.NewLine + info.Stats.UsersCount.ToString + "/" + info.Stats.CheckinsCount.ToString + Environment.NewLine + info.Location.Address + Environment.NewLine + info.Location.City + info.Location.State + Environment.NewLine + info.Location.Latitude.ToString + Environment.NewLine + info.Location.Longitude.ToString
        End Get
    End Property
    Public Function GetContent(ByVal method As String, _
                ByVal requestUri As Uri, _
                ByVal param As Dictionary(Of String, String), _
                ByRef content As String) As HttpStatusCode

        Dim webReq As HttpWebRequest = CreateRequest(method, _
                                                    requestUri, _
                                                    param, _
                                                    False)
        Dim code As HttpStatusCode
        code = GetResponse(webReq, content, Nothing, False)
        Return code
    End Function

#Region "FourSquare DataModel"

    Public Class FourSquareDataModel

        <DataContract()>
        Public Class FourSquareData
            <DataMember(Name:="meta", isRequired:=False)> Public Meta As Meta
            <DataMember(Name:="response", isRequired:=False)> Public Response As Response
        End Class

        <DataContract()>
        Public Class Response
            <DataMember(Name:="venue", isRequired:=False)> Public Venue As Venue
        End Class

        <DataContract()>
        Public Class Venue
            <DataMember(Name:="id")> Public Id As String
            <DataMember(Name:="name")> Public Name As String
            <DataMember(Name:="location")> Public Location As Location
            <DataMember(Name:="verified")> Public Verified As Boolean
            <DataMember(Name:="stats")> Public Stats As Stats
            <DataMember(Name:="mayor")> Public Mayor As Mayor
            <DataMember(Name:="shortUrl")> Public ShortUrl As String
            <DataMember(Name:="timeZone")> Public TimeZone As String
        End Class

        <DataContract()>
        Public Class Location
            <DataMember(Name:="address")> Public Address As String
            <DataMember(Name:="city")> Public City As String
            <DataMember(Name:="state")> Public State As String
            <DataMember(Name:="lat")> Public Latitude As Double
            <DataMember(Name:="lng")> Public Longitude As Double
        End Class

        <DataContract()>
        Public Class Stats
            <DataMember(Name:="checkinsCount")> Public CheckinsCount As Integer
            <DataMember(Name:="usersCount")> Public UsersCount As Integer
        End Class


        <DataContract()>
        Public Class Mayor
            <DataMember(Name:="count")> Public Count As Integer
            <DataMember(Name:="user", isrequired:=False)> Public User As FoursquareUser
        End Class

        <DataContract()>
        Public Class FoursquareUser
            <DataMember(Name:="id")> Public Id As Integer
            <DataMember(Name:="firstName")> Public FirstName As String
            <DataMember(Name:="photo")> Public Photo As String
            <DataMember(Name:="gender")> Public Gender As String
            <DataMember(Name:="homeCity")> Public HomeCity As String
        End Class

        <DataContract()>
        Public Class Meta
            <DataMember(Name:="code")> Public Code As Integer
            <DataMember(Name:="errorType", IsRequired:=False)> Public ErrorType As String
            <DataMember(Name:="errorDetail", IsRequired:=False)> Public ErrorDetail As String
        End Class
    End Class
#End Region


End Class

File: TweenCS/Foursquare.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Net;
using System.Text.RegularExpressions;
using System.Runtime.Serialization;

namespace Tween
{
    public class Foursquare : HttpConnection
    {
        private static Foursquare _instance = new Foursquare();
        public static Foursquare GetInstance
        {
            get
            {
                return _instance;
            }
        }

        private Dictionary<string, string> _authKey = new Dictionary<string, string>
        {
            {"client_id", "5H3K5YQPT55DNQUFEOAJFNJA5D01ZJGO2ITEAJ3ASRIDONUB"},
            {"client_secret", "JFRHP1L451M3AEPF11UZLTIIUZCZTZRVHVOWB5TQ0AJOVPBB"},
        };

        private Dictionary<string, Google.GlobalLocation> CheckInUrlsVenueCollection = new Dictionary<string, Google.GlobalLocation>();

        private string GetVenueId(string url)
        {
            var content = "";
            try
            {
                var res = GetContent("GET", new Uri(url), null, ref content);
                if (res != HttpStatusCode.OK) return "";
            }
            catch (Exception)
            {
                return "";
            }
            var mc = Regex.Match(content, "/venue/(?<venueId>[0-9]+)", RegexOptions.IgnoreCase);
            if (mc.Success)
            {
                var vId = mc.Result("${venueId}");
                return vId;
            }
            else
            {
                return "";
            }
        }

        private FourSquareDataModel.Venue GetVenueInfo(string venueId)
        {
            var content = "";
            try
            {
                var res = GetContent("GET",
                                     new Uri("https://api.foursquare.com/v2/venues/" + venueId),
                                     _authKey,
                                     ref content);

                if (res == HttpStatusCode.OK)
                {
                    FourSquareDataModel.FourSquareData curData = null;
                    try
                    {
                        curData = MyCommon.CreateDataFromJson<FourSquareDataModel.FourSquareData>(content);
                    }
                    catch (Exception)
                    {
                        return null;
                    }

                    return curData.Response.Venue;
                }
                else
                {
                    //FourSquareDataModel.FourSquareData curData = null;
                    //try
                    //{
                    //    curData = CreateDataFromJson<FourSquareDataModel.FourSquareData>(content);
                    //}
                    //catch (Exception)
                    //{
                    //    return null;;
                    //}
                    //MessageBox.Show(res.ToString() + Environment.NewLine + curData.Meta.ErrorType + Environment.NewLine + curData.Meta.ErrorDetail);
                    return null;
                }
            }
            catch (Exception)
            {
                return null;
            }
        }

        public string GetMapsUri(string url, ref string refText)
        {
            if (!AppendSettingDialog.Instance.IsPreviewFoursquare) return null;

            var urlId = Regex.Replace(url, @"https?://(4sq|foursquare)\.com/", "");

            if (CheckInUrlsVenueCollection.ContainsKey(urlId))
            {
                refText = CheckInUrlsVenueCollection[urlId].LocateInfo;
                return new Google().CreateGoogleStaticMapsUri(CheckInUrlsVenueCollection[urlId]);
            }

            FourSquareDataModel.Venue curVenue = null;
            var venueId = GetVenueId(url);
            if (string.IsNullOrEmpty(venueId)) return null;

            curVenue = GetVenueInfo(venueId);
            if (curVenue == null) return null;

            var curLocation = new Google.GlobalLocation {Latitude = curVenue.Location.Latitude, Longitude = curVenue.Location.Longitude, LocateInfo = CreateVenueInfoText(curVenue)};
            //例外発生の場合があるため
            if (!CheckInUrlsVenueCollection.ContainsKey(urlId)) CheckInUrlsVenueCollection.Add(urlId, curLocation);
            refText = curLocation.LocateInfo;
            return new Google().CreateGoogleStaticMapsUri(curLocation);
        }

        private string CreateVenueInfoText(FourSquareDataModel.Venue info)
        {
            return info.Name + Environment.NewLine + info.Stats.UsersCount.ToString() + "/" + info.Stats.CheckinsCount.ToString() + Environment.NewLine + info.Location.Address + Environment.NewLine + info.Location.City + info.Location.State + Environment.NewLine + info.Location.Latitude.ToString() + Environment.NewLine + info.Location.Longitude.ToString();
        }
        public HttpStatusCode GetContent(string method,
                    Uri requestUri,
                    Dictionary<string, string> param,
                    ref string content)
        {
            var webReq  = CreateRequest(method,
                                        requestUri,
                                        param,
                                        false);
            var code = GetResponse(webReq, out content, null, false);
            return code;
        }

    #region "FourSquare DataModel"

        public class FourSquareDataModel
        {
            [DataContract]
            public class FourSquareData
            {
                [DataMember(Name = "meta", IsRequired = false)] public Meta Meta;
                [DataMember(Name = "response", IsRequired = false)] public Response Response;
            }

            [DataContract]
            public class Response
            {
                [DataMember(Name = "venue", IsRequired = false)] public Venue Venue;
            }

            [DataContract]
            public class Venue
            {
                [DataMember(Name = "id")] public string Id;
                [DataMember(Name = "name")] public string Name;
                [DataMember(Name = "location")] public Location Location;
                [DataMember(Name = "verified")] public bool Verified;
                [DataMember(Name = "stats")] public Stats Stats;
                [DataMember(Name = "mayor")] public Mayor Mayor;
                [DataMember(Name = "shortUrl")] public string ShortUrl;
                [DataMember(Name = "timeZone")] public string TimeZone;
            }

            [DataContract]
            public class Location
            {
                [DataMember(Name = "address")] public string Address;
                [DataMember(Name = "city")] public string City;
                [DataMember(Name = "state")] public string State;
                [DataMember(Name = "lat")] public double Latitude;
                [DataMember(Name = "lng")] public double Longitude;
            }

            [DataContract]
            public class Stats
            {
                [DataMember(Name = "checkinsCount")] public int CheckinsCount;
                [DataMember(Name = "usersCount")] public int UsersCount;
            }


            [DataContract]
            public class Mayor
            {
                [DataMember(Name = "count")] public int Count;
                [DataMember(Name = "user", IsRequired = false)] public FoursquareUser User;
            }

            [DataContract]
            public class FoursquareUser
            {
                [DataMember(Name = "id")] public int Id;
                [DataMember(Name = "firstName")] public string FirstName;
                [DataMember(Name = "photo")] public string Photo;
                [DataMember(Name = "gender")] public string Gender;
                [DataMember(Name = "homeCity")] public string HomeCity;
            }

            [DataContract]
            public class Meta
            {
                [DataMember(Name = "code")] public int Code;
                [DataMember(Name = "errorType", IsRequired = false)] public string ErrorType;
                [DataMember(Name = "errorDetail", IsRequired = false)] public string ErrorDetail;
            }
        }
    #endregion


    }
}
